runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -8077,75 +8106,107 @@ static TR::Register *genCAS(TR::Node *node, TR::CodeGenerator *cg, TR::Register
    TR::LabelSymbol *loopLabel = generateLabelSymbol(cg);
    generateLabelInstruction(cg, TR::InstOpCode::label, node, loopLabel);
 
-   generateTrg1MemInstruction(cg, isLong ? TR::InstOpCode::ldarx : TR::InstOpCode::lwarx, node, resultReg, TR::MemoryReference::createWithIndexReg(cg, objReg, offsetReg, isLong ? 8 : 4));
+   generateTrg1MemInstruction(cg, reservedLoadOpCode, node, resultReg, TR::MemoryReference::createWithIndexReg(cg, objReg, offsetReg, dataSize));
    if (oldValueInReg)
-      generateTrg1Src2Instruction(cg, isLong ? TR::InstOpCode::cmp8 : TR::InstOpCode::cmp4, node, cndReg, resultReg, oldVReg);
+      generateTrg1Src2Instruction(cg, compareOpCode, node, cndReg, resultReg, oldVReg);
    else
-      generateTrg1Src1ImmInstruction(cg, isLong ? TR::InstOpCode::cmpi8 : TR::InstOpCode::cmpi4, node, cndReg, resultReg, oldValue);
-   generateTrg1ImmInstruction(cg, TR::InstOpCode::li, node, resultReg, 0);
+      generateTrg1Src1ImmInstruction(cg, compareImmOpCode, node, cndReg, resultReg, oldValue);
+
+   if (!isExchange)
+      {
+      generateTrg1ImmInstruction(cg, TR::InstOpCode::li, node, resultReg, 0);
+      }
+   else if (isReference && comp->target().is64Bit() && comp->useCompressedPointers())
+      {
+      genDecompressPointer(cg, node, resultReg);
+      }
 
```

Comment:
i can see some contradictions here: you seemed assuming that the coming-in old/new values in registers are still compressed (if we run in compressedRef mode of course).  you are using them as if the assumption were true. however, compressed values should never exist for any substantial duration (e.g. over a GC point). usually, they are decompressed right away after loading, as you did above for `resultReg`.   that is the contradiction i referred to (i.e. old/new values passed in registers -- the nodes might have existed there and re-referenced many times for a long time -- are still compressed).

Reply:
The reason for this inconsistency is due to this:
https://github.com/IBMJimmyk/openj9/blob/48f927706ac45e937e059a2879d2007421e4ef24/runtime/compiler/codegen/J9CodeGenerator.cpp#L655-L677

There is existing logic that causes the compressedrefs to not be decompressed before being used by codegen. I had planned to get rid of this after looking at Z since it affects that platform as well. But, I could try removing it just for Power if you think that would be better.

Reply:
i am ok with visiting it later after this merge.